### PR TITLE
crates/sound: Rebase pipewire backend

### PR DESCRIPTION
**Pull request by dorindabassey**
Rebase pipewire Backend

### CODE REVIEW

File: crates/sound/src/audio_backends/pipewire.rs [hunk 1]
```diff
@@ -1,30 +1,566 @@
...
+        let format_str = format!($($args)*);
+        let format_cstr = CString::new(format_str).unwrap();
+        unsafe {
+            pw::sys::pw_properties_set(
+                $properties.as_ptr(),
+                key_cstr.as_ptr(),
+                format_cstr.as_ptr(),
+            )
+        }
+    }};
```
Comment (bilelmoussaoui) [inline crates/sound/src/audio_backends/pipewire.rs]: Any reason of using a macro here instead of function?  maybe this should be also upstreamed and be part of `pipewire::Properties`

File: crates/sound/src/audio_backends/pipewire.rs [hunk 2]
```diff
@@ -1,30 +1,566 @@
...
+        if inner.is_null() {
+            None
+        } else {
+            Some(Self(
+                NonNull::new(inner).expect("pw_thread_loop can't be null"),
+            ))
+        }
+    }
+
+    pub fn get_loop(&self) -> PwThreadLoopTheLoop {
```
Comment (bilelmoussaoui) [inline crates/sound/src/audio_backends/pipewire.rs]: the name of this struct is pretty bad, i initially wrote that just to give you an idea of how to make the loop would work

File: crates/sound/src/audio_backends/pipewire.rs [hunk 3]
```diff
@@ -1,30 +1,566 @@
...
+        unsafe { &*(self.0.as_ptr() as *mut LoopRef) }
+    }
+}
+
+// SAFETY: Safe as the structure can be sent to another thread.
+unsafe impl Send for PwBackend {}
+
+// SAFETY: Safe as the structure can be shared with another thread as the state
+// is protected with a lock.
+unsafe impl Sync for PwBackend {}
```
Comment (bilelmoussaoui) [inline crates/sound/src/audio_backends/pipewire.rs:138]:
```suggestion
unsafe impl Sync for PwBackend {}

```

File: crates/sound/src/audio_backends/pipewire.rs [hunk 4]
```diff
@@ -1,30 +1,566 @@
...
+                if id == PW_ID_CORE && seq == pending {
+                    thread_clone.signal();
+                }
+            })
+            .register();
+
+        thread_loop.wait();
+        thread_loop.unlock();
+
+        println!("pipewire backend running");
```
Comment (bilelmoussaoui) [inline crates/sound/src/audio_backends/pipewire.rs]: use the log crate instead

File: crates/sound/src/audio_backends/pipewire.rs [hunk 5]
```diff
@@ -1,30 +1,566 @@
...
+        let desc_request = &descriptors[0];
+        let request = msg
+            .desc_chain
+            .memory()
+            .read_obj::<VirtioSndPcmSetParams>(desc_request.addr())
+            .unwrap();
+        {
+            let stream_clone = self.stream_params.clone();
+            let mut stream_params = stream_clone.write().unwrap();
+            let st = &mut stream_params[stream_id as usize];
```
Comment (bilelmoussaoui) [inline crates/sound/src/audio_backends/pipewire.rs]: use get_mut instead

File: crates/sound/src/audio_backends/pipewire.rs [hunk 6]
```diff
@@ -1,30 +1,566 @@
...
+
+        Ok(())
+    }
+
+    fn start(&self, stream_id: u32) -> Result<()> {
+        debug!("pipewire start");
+        let start_result = self.stream_params.write().unwrap()[stream_id as usize]
+            .state
+            .start();
+        if let Err(err) = start_result {
```
Comment (bilelmoussaoui) [inline crates/sound/src/audio_backends/pipewire.rs:514]: this looks wrong, as you are not propagating the error unless that is intended

File: crates/sound/src/audio_backends/pipewire.rs [hunk 7]
```diff
@@ -1,30 +1,566 @@
...
+        }
+        Ok(())
+    }
+
+    fn stop(&self, stream_id: u32) -> Result<()> {
+        debug!("pipewire stop");
+        let stop_result = self.stream_params.write().unwrap()[stream_id as usize]
+            .state
+            .stop();
+        if let Err(err) = stop_result {
```
Comment (bilelmoussaoui) [inline crates/sound/src/audio_backends/pipewire.rs:534]: same as start

File: crates/sound/src/audio_backends/pipewire.rs [hunk 8]
```diff
@@ -1,30 +1,566 @@
...
+    }
+
+    pub fn wait(&self) {
+        unsafe {
+            pw_thread_loop_wait(self.0.as_ptr());
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
```
Comment (bilelmoussaoui) [inline crates/sound/src/audio_backends/pipewire.rs:114]: Clone is wrong here as well, you need to wrap the inner pw_loop with either Arc or Rc.

File: crates/sound/src/audio_backends/pipewire.rs [hunk 9]
```diff
@@ -1,30 +1,566 @@
...
+    collections::HashMap,
+    convert::TryInto,
+    ops::Deref,
+    ptr,
+    ptr::NonNull,
+    sync::{Arc, RwLock},
+};
 
+use log::debug;
+use pipewire as pw;
```
Comment (bilelmoussaoui) [inline crates/sound/src/audio_backends/pipewire.rs]: You should rename the crate in Cargo.toml

File: crates/sound/src/audio_backends/pipewire.rs [hunk 10]
```diff
@@ -1,30 +1,566 @@
...
+use pw::spa::param::ParamType;
+use pw::spa::pod::Pod;
```
Comment (bilelmoussaoui) [inline crates/sound/src/audio_backends/pipewire.rs]: group these imports please

File: crates/sound/Cargo.toml [hunk 11]
```diff
@@ -13,18 +13,15 @@ edition = "2018"
 default = ["null-backend", "alsa-backend", "pw-backend"]
 null-backend = []
 alsa-backend = ["dep:alsa"]
-pw-backend = ["pipewire", "libspa", "pipewire-sys", "libspa-sys", "bindgen"]
+pw-backend = ["pipewire", "bindgen"]
 
 [dependencies]
 alsa = { version = "0.7", optional = true }
-bindgen = { version = "0.64.0", optional = true }
+bindgen = { version = "0.66.0", optional = true }
```
Comment (bilelmoussaoui) [inline crates/sound/Cargo.toml]: do you really need bindgen here? i don't think so

**Review by bilelmoussaoui — COMMENTED** [review]
random comments; will do a proper review later

File: crates/sound/src/audio_backends/pipewire.rs [hunk 12]
```diff
@@ -1,30 +1,566 @@
...
+                        let datas = (*buf).datas;
+
+                        let p = (*datas).data as *mut u8;
+
+                        if p.is_null() {
+                            return;
+                        }
+
+                        // to calculate as sizeof(int16_t) * NR_CHANNELS
+                        let frame_size = info.channels as u32 * size_of::<u16>() as u32;
```
Comment (bilelmoussaoui) [inline crates/sound/src/audio_backends/pipewire.rs]:
```suggestion
                        // to calculate as sizeof(int16_t) * NR_CHANNELS
                        let frame_size = info.channels as u32 * size_of::<i16>() as u32;
```
no?

File: crates/sound/src/audio_backends/pipewire.rs [hunk 13]
```diff
@@ -1,30 +1,566 @@
...
+                        let req = (*b).requested * (frame_size as u64);
+                        let mut n_bytes = cmp::min(req as u32, (*datas).maxsize);
+
+                        let stream2 = &mut streams.write().unwrap()[stream_id as usize];
```
Comment (bilelmoussaoui) [inline crates/sound/src/audio_backends/pipewire.rs]: use get_mut instead

File: crates/sound/src/audio_backends/pipewire.rs [hunk 14]
```diff
@@ -1,30 +1,566 @@
...
+            let mut stream_hash = self.stream_hash.write().unwrap();
+            let mut stream_listener = self.stream_listener.write().unwrap();
+            let st_buffer = &mut self.stream_params.write().unwrap();
+
+            if let Some(stream) = stream_hash.get(&stream_id) {
+                stream.disconnect().expect("could not disconnect stream");
+                std::mem::take(&mut st_buffer[stream_id as usize].buffers);
+            } else {
+                return Err(Error::StreamWithIdNotFound(stream_id));
+            };
```
Comment (bilelmoussaoui) [inline crates/sound/src/audio_backends/pipewire.rs]:
```suggestion
let Some(stream) = stream_hash.get(&stream_id) else {
    return Err(Error::StreamWithIdNotFound(stream_id));
};
stream.disconnect().expect("could not disconnect stream");
std::mem::take(&mut st_buffer[stream_id as usize].buffers);
```

Comment (bilelmoussaoui) [reply crates/sound/src/audio_backends/pipewire.rs]: same elsewhere

File: crates/sound/src/audio_backends/pipewire.rs (hunk 1 above)
Comment (dorindabassey) [reply crates/sound/src/audio_backends/pipewire.rs]: I wanted to upstream this macro but the pipewire-rs upstream maintainer would not accept it, and suggested instead for users to set the format macro themselves in their projects.

Comment (bilelmoussaoui) [reply crates/sound/src/audio_backends/pipewire.rs]: This should really be a function imho.

File: crates/sound/src/audio_backends/pipewire.rs (hunk 2 above)
Comment (dorindabassey) [reply crates/sound/src/audio_backends/pipewire.rs]: yes, I should probably change the struct name. thanks.

File: crates/sound/src/audio_backends/pipewire.rs (hunk 5 above)
Comment (dorindabassey) [reply crates/sound/src/audio_backends/pipewire.rs]: is there any significant difference between using &mut and get_mut()?

File: crates/sound/src/audio_backends/pipewire.rs (hunk 9 above)
Comment (dorindabassey) [reply crates/sound/src/audio_backends/pipewire.rs]: it's easier to refer to pipewire using the shorter name `pw` in the current scope. wdym by to rename the crate in cargo.toml?

File: crates/sound/Cargo.toml (hunk 11 above)
Comment (dorindabassey) [reply crates/sound/Cargo.toml]: I forgot how that landed in there, yes i should remove it

File: crates/sound/src/audio_backends/pipewire.rs (hunk 1 above)
Comment (dorindabassey) [reply crates/sound/src/audio_backends/pipewire.rs]: given that this involves variadic argument I think macro_rules! handle it better.

File: crates/sound/src/audio_backends/pipewire.rs (hunk 6 above)
Comment (dorindabassey) [reply crates/sound/src/audio_backends/pipewire.rs:514]: it's intended, we don't want it to panic in that case, logging the error should be enough.

File: crates/sound/src/audio_backends/pipewire.rs (hunk 9 above)
Comment (bilelmoussaoui) [reply crates/sound/src/audio_backends/pipewire.rs]: `pw = { package = "pipewire", ....}`  and you can drop that line

File: crates/sound/src/audio_backends/pipewire.rs (hunk 6 above)
Comment (bilelmoussaoui) [reply crates/sound/src/audio_backends/pipewire.rs:514]: Aha, that is the second time i comment about the same thing. Definitely worth a comment then :)

File: crates/sound/src/audio_backends/pipewire.rs (hunk 5 above)
Comment (bilelmoussaoui) [reply crates/sound/src/audio_backends/pipewire.rs]: Actually, my poor brain thought that only `std::ops::Index` existed and not `std::ops::IndexMut`. but the main difference is using `[index]` would panic if the index is out of bounds, where `get_mut` would return `Option<&mut T>` which might be useful here to at least log a proper "failure error" with `get_mut(stream_id as usize).expect("Stream must exist to set it parameters")`

File: crates/sound/src/audio_backends/pipewire.rs [hunk 15]
```diff
@@ -1,30 +1,566 @@
...
+
+            let mut param = [Pod::from_bytes(&values).unwrap()];
+
+            let props = properties! {
+                *pw::keys::MEDIA_TYPE => "Audio",
+                *pw::keys::MEDIA_CATEGORY => "Playback",
+            };
+            // 75% of the timer period for faster updates
+            // Todo: investigate the right value to use
+            let buf_samples: u32 = info.rate * 3 / 4 / 100;
```
Comment (MatiasVara) [inline crates/sound/src/audio_backends/pipewire.rs]: I am not completely sure if we need to set the latency. Maybe we can drop it here and just create an issue to deal with that later.

File: crates/sound/src/audio_backends/pipewire.rs [hunk 16]
```diff
@@ -1,30 +1,566 @@
...
+                            return;
+                        }
+
+                        // to calculate as sizeof(int16_t) * NR_CHANNELS
+                        let frame_size = info.channels as u32 * size_of::<u16>() as u32;
+                        let req = (*b).requested * (frame_size as u64);
+                        let mut n_bytes = cmp::min(req as u32, (*datas).maxsize);
+
+                        let stream2 = &mut streams.write().unwrap()[stream_id as usize];
+                        let req2 = stream2.buffers.front_mut().unwrap();
```
Comment (MatiasVara) [inline crates/sound/src/audio_backends/pipewire.rs]: We have to handle the case in which req2 is none because there is no new buffer. In that case, we can just fill it with silence by using something like audio_pcm_info_clear_buf().

File: crates/sound/src/audio_backends/pipewire.rs (hunk 1 above)
Comment (bilelmoussaoui) [reply crates/sound/src/audio_backends/pipewire.rs]: but even the C function that you are wrapping here is not using a variadic argument, doing so in rust just hurts readability and macros are terrible for LSPs (language servers providers like rust analyzer)

File: crates/sound/src/audio_backends/pipewire.rs (hunk 5 above)
Comment (dorindabassey) [reply crates/sound/src/audio_backends/pipewire.rs]: ok make sense, thanks

File: crates/sound/src/audio_backends/pipewire.rs (hunk 6 above)
Comment (dorindabassey) [reply crates/sound/src/audio_backends/pipewire.rs:514]: noted

File: crates/sound/src/audio_backends/pipewire.rs (hunk 15 above)
Comment (dorindabassey) [reply crates/sound/src/audio_backends/pipewire.rs]: Ok, I agree with you on that. if need be we set the latency.

File: crates/sound/src/audio_backends/pipewire.rs (hunk 16 above)
Comment (dorindabassey) [reply crates/sound/src/audio_backends/pipewire.rs]: do you mean to complete the todo in `if avail <= 0 {}` pipewire.rs:L440

File: crates/sound/src/audio_backends/pipewire.rs (hunk 8 above)
Comment (dorindabassey) [reply crates/sound/src/audio_backends/pipewire.rs:114]: I found it a bit problematic wrapping the inner pw_loop with Arc or Rc, I would leave it with clone, but what is wrong about using clone in this case ? :thinking:

File: crates/sound/src/audio_backends/pipewire.rs (hunk 16 above)
Comment (MatiasVara) [reply crates/sound/src/audio_backends/pipewire.rs]: Not exactly, I am referring to the case in which `front_mut()` returns "none". We should have something like:
```rust
   let buf = match stream_reqs.buffers.front_mut() {
        Some(x) => x,
        None => return // I think in this case we have to do something similar to the case avail == 0
    };
```

Comment (dorindabassey) [reply crates/sound/src/audio_backends/pipewire.rs]: Ah I see what you mean, ok ok.

File: crates/sound/src/audio_backends/pipewire.rs (hunk 8 above)
Comment (bilelmoussaoui) [reply crates/sound/src/audio_backends/pipewire.rs:114]: > I found it a bit problematic wrapping the inner pw_loop with Arc or Rc,

Which part is problematic ? Because `pw_loop` is not reference counted, see how pipewire-rs handles it https://pipewire.pages.freedesktop.org/pipewire-rs/src/pipewire/main_loop.rs.html#13

File: crates/sound/src/lib.rs [hunk 17]
```diff
@@ -77,6 +77,10 @@ pub enum Error {
     UnexpectedDescriptorSize(usize, u32),
     #[error("Protocol or device error: {0}")]
     Stream(stream::Error),
+    #[error("Stream with id {0} not found")]
+    StreamWithIdNotFound(u32),
+    #[error("Channel number not supported")]
+    ChannelNotSupported,
```
Comment (bilelmoussaoui) [inline crates/sound/src/lib.rs]: Maybe keep the information about which channel is not supported?

File: crates/sound/src/stream.rs [hunk 18]
```diff
@@ -8,6 +8,7 @@ use vm_memory::{Le32, Le64};
 
 use crate::{virtio_sound::*, IOMessage, SUPPORTED_FORMATS, SUPPORTED_RATES};
 
+pub const NR_STREAMS: usize = 1;
```
Comment (bilelmoussaoui) [inline crates/sound/src/stream.rs]: Any reason this was added here if it is only used by the pipewire backend?

File: crates/sound/src/device.rs [hunk 19]
```diff
@@ -397,6 +397,10 @@ impl VhostUserSoundThread {
 
         let mut stream_ids = BTreeSet::default();
 
+        // temporary workaround for processing the buffers speed
```
Comment (bilelmoussaoui) [inline crates/sound/src/device.rs]: This should be set in a separate commit for historical purposes & it can then be reverted cleanly once the issue is resolved

File: crates/sound/src/audio_backends/pipewire.rs [hunk 20]
```diff
@@ -1,30 +1,542 @@
...
+    ops::Deref,
+    ptr,
+    ptr::NonNull,
+    sync::{Arc, RwLock},
+};
+
+use log::debug;
+use pw::spa::{param::ParamType, pod::Pod};
+
+use pw::spa::sys::{
```
Comment (bilelmoussaoui) [inline crates/sound/src/audio_backends/pipewire.rs]: you already do use `pw::spa` so you can drop the `pw` prefix here
```suggestion
use spa::sys::{
```

File: crates/sound/src/audio_backends/pipewire.rs [hunk 21]
```diff
@@ -1,30 +1,542 @@
...
+    convert::TryInto,
+    mem::size_of,
+    ops::Deref,
+    ptr,
+    ptr::NonNull,
+    sync::{Arc, RwLock},
+};
+
+use log::debug;
+use pw::spa::{param::ParamType, pod::Pod};
```
Comment (bilelmoussaoui) [inline crates/sound/src/audio_backends/pipewire.rs]:
```suggestion
use spa::{param::ParamType, pod::Pod};
```

File: crates/sound/src/audio_backends/pipewire.rs [hunk 22]
```diff
@@ -1,30 +1,542 @@
...
+                channels: params.channels as u32,
+                position: pos,
+            };
+
+            let mut audio_info = spa::param::audio::AudioInfoRaw::new();
+            audio_info.set_format(spa::param::audio::AudioFormat::S16LE);
+            audio_info.set_rate(info.rate);
+            audio_info.set_channels(info.channels);
+
+            let values: Vec<u8> = pw::spa::pod::serialize::PodSerializer::serialize(
```
Comment (bilelmoussaoui) [inline crates/sound/src/audio_backends/pipewire.rs]:
```suggestion
            let values: Vec<u8> = spa::pod::serialize::PodSerializer::serialize(
```
to be consistent with the lines above & similarly elsewhere

File: crates/sound/src/audio_backends/pipewire.rs [hunk 23]
```diff
@@ -1,30 +1,542 @@
...
+                std::io::Cursor::new(Vec::new()),
+                &pw::spa::pod::Value::Object(pw::spa::pod::Object {
+                    type_: pw::spa::sys::SPA_TYPE_OBJECT_Format,
```
Comment (bilelmoussaoui) [inline crates/sound/src/audio_backends/pipewire.rs]: I find it weird that certain ffi (sys) types are explicitly included and others are not.

File: crates/sound/src/audio_backends/pipewire.rs [hunk 24]
```diff
@@ -1,30 +1,542 @@
...
+            };
+
+            let stream = pw::stream::Stream::new(&self.core, "audio-output", props)
+                .expect("could not create new stream");
+
+            let streams = self.stream_params.clone();
+
+            let listener_stream = stream
+                .add_local_listener()
+                .state_changed(move |old, new| {
```
Comment (bilelmoussaoui) [inline crates/sound/src/audio_backends/pipewire.rs]:
```suggestion
                .state_changed(|old, new| {
```
there is nothing moved inside the closure

File: crates/sound/src/audio_backends/pipewire.rs [hunk 25]
```diff
@@ -1,30 +1,542 @@
...
+        thread_loop.unlock();
+
+        log::trace!("pipewire backend running");
+
+        Self {
+            stream_params,
+            thread_loop,
+            core,
+            context,
+            stream_hash: RwLock::new(HashMap::with_capacity(NR_STREAMS)),
```
Comment (MatiasVara) [inline crates/sound/src/audio_backends/pipewire.rs]: I think the number of streams can be got from `stream_params.len()` thus NR_STREAMS can be removed.

File: crates/sound/src/audio_backends/pipewire.rs [hunk 26]
```diff
@@ -1,30 +1,542 @@
...
+                            return;
+                        };
+
+                        let mut start = buffer.pos;
+
+                        let avail = (buffer.bytes.len() - start) as i32;
+
+                        if avail <= 0 {
+                            // pad with silence
+                            ptr::write_bytes(p, 0, n_bytes as usize);
```
Comment (MatiasVara) [inline crates/sound/src/audio_backends/pipewire.rs:424]: is this the correct function to build silent?

File: crates/sound/src/audio_backends/pipewire.rs (hunk 8 above)
Comment (dorindabassey) [reply crates/sound/src/audio_backends/pipewire.rs:114]: Here, even after wrapping in Rc they still use clone. I can't avoid using clone. which is why i want to understand why using clone here is wrong.

File: crates/sound/src/stream.rs (hunk 18 above)
Comment (MatiasVara) [reply crates/sound/src/stream.rs]: This can be removed.

File: crates/sound/src/audio_backends/pipewire.rs (hunk 26 above)
Comment (dorindabassey) [reply crates/sound/src/audio_backends/pipewire.rs:424]: in pipewire C API they use memset for silence, so i found the equivalent in rust to be this, Which I presume is correct.

File: crates/sound/src/device.rs (hunk 19 above)
Comment (dorindabassey) [reply crates/sound/src/device.rs]: noted

File: crates/sound/src/audio_backends/pipewire.rs (hunk 8 above)
Comment (bilelmoussaoui) [reply crates/sound/src/audio_backends/pipewire.rs:114]: You should keep clone yes, but only after you wrapped the inner type with Rc/Arc

**Review by MatiasVara — APPROVED** [review]

File: crates/sound/src/audio_backends/pipewire.rs [hunk 27]
```diff
@@ -1,30 +1,548 @@
...
+
+    pub fn wait(&self) {
+        unsafe {
+            pw_thread_loop_wait(self.0.as_ptr());
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+struct PwGetLoop {
```
Comment (bilelmoussaoui) [inline crates/sound/src/audio_backends/pipewire.rs]:
```suggestion
struct PwInnerLoop {
```
otherwise lgtm

**Review by bilelmoussaoui — APPROVED** [review]
